package/MDAnalysis/analysis/hbonds/hbond_analysis.py

```diff
@@ -816,20 +816,21 @@ def _update_selection_2(self):
         self._s2_donors = {}
         self._s2_donors_h = {}
         self._s2_acceptors = {}
-        if self.selection1_type in ('donor', 'both'):
-            self._s2_acceptors = self._s2.select_atoms(
-                'name {0}'.format(' '.join(self.acceptors)))
-            self.logger_debug("Selection 2 acceptors: {0:d}".format(len(self._s2_acceptors)))
-        if self.selection1_type in ('acceptor', 'both'):
-            self._s2_donors = self._s2.select_atoms(
-                'name {0}'.format(' '.join(self.donors)))
-            self._s2_donors_h = {}
-            for i, d in enumerate(self._s2_donors):
-                tmp = self._get_bonded_hydrogens(d)
-                if tmp:
-                    self._s2_donors_h[i] = tmp
-            self.logger_debug("Selection 2 donors: {0:d}".format(len(self._s2_donors)))
-            self.logger_debug("Selection 2 donor hydrogens: {0:d}".format(len(self._s2_donors_h)))
+        if self._s2:
```

Comment:
In my opinion, it would be cleaner to just return here instead of doing a big `if` block. If you re-order the code so that all data structures are initialized and then use the `if not self._s2` where the logger is already reporting  this problem and `return` from there.

```python
         self._s2_donors = {}
         self._s2_donors_h = {}
         self._s2_acceptors = {}

         if not self._s2:
            # nothing to do here, just keep empty data structures
            logger.warn('Selection 2 "{0}" did not select any atoms.'.format(
                 str(self.selection2)[:80]))
            return

        if self.selection1_type in ('donor', 'both'):
            self._s2_acceptors = self._s2.select_atoms(...)
            ...
```

This is reasonably clean (and the diff will also be smaller).